lib/quasi-chainmail.js

```diff
@@ -65,14 +65,16 @@ enumDecl <- ENUM type LBRACE (IDENT _SEMI)* RBRACE
 STRUCT <- "struct" _WS;
 structDecl <- STRUCT type LBRACE memberDecl* RBRACE
   ${(_, stype, _2, members, _3) =>
-  ['struct', stype[1], stype[2], members]};
+  ['struct', stype, members]};
 
 memberDecl <- paramDecl _SEMI / typeDecl;
 
 INTERFACE <- "interface" _WS;
 interfaceDecl <- INTERFACE type extends? LBRACE methodDecl* RBRACE
   ${(_, itype, ext, _2, methods, _3) =>
-  ['interface', itype, deopt(ext), methods]};
+   ext.length === 0
+     ? ['interface', itype, methods]
+     : ['interface', itype, [ 'extends', ext[0]], methods]};
 
```

Comment:
It's not good practice to have the same AST node (tagged tuple) have different arity.  The only time we can get away with it is when the later members are actually `undefined`.  So, I'd rather see this remain the same unless you want to introduce two different nodes rather than just `interface`.

You can think of an AST node as a reified function call (`[functionName, ...arguments]`).  If you change the type of an argument between calls to the same function, you're needing to discriminate its type on more than just the `functionName`.  Visitors then become a bit of a mess.

Reply:
Thanks. That's the feedback I needed. I'm happy to leave it as it was, though the result is a bit noisy. Soon enough, the output won't be something we read with the unaided eye.